chronos/streams/tlsstream.nim

```diff
@@ -465,8 +470,8 @@ proc newTLSClientAsyncStream*(rsource: AsyncStreamReader,
     sslEngineSetX509(res.ccontext.eng, addr res.xwc.vtable)
   else:
     sslClientInitFull(res.ccontext, addr res.x509,
-                      unsafeAddr MozillaTrustAnchors[0],
-                      uint(len(MozillaTrustAnchors)))
+                      unsafeAddr trustAnchors[0],
```

Comment:
I'm not sure about the memory safety of this, seems like the caller could discard the anchors after calling this provider, and it will then lead to garbage memory

Reply:
Totally agree.
First of all you should add assertion check at the beginning of this procedure.
```nim
  doAssert(len(trustAnchors) > 0, "Some comment")
```

Reply:
Would changing the parameter to `trustAnchors: sink openArray[X509TrustAnchor] = MozillaTrustAnchors` be sufficient? I'm not terribly familiar with how move/sink works, but my gut says this situation is what a `sink` is for.

This simple test program below seems to confirm that, but:

1. I don't know if it just gets lucky that the array's memory happens to still be intact
2. Or if it copies the whole contents of the array, which might be too expensive for `newTLSClientAsyncStream` to have to do each time.

```nim
type
  Thing = object
    name: string

var globalThings: ptr Thing
var globalThingsLen: int

proc save(arr: sink openArray[Thing]) =
  globalThings = unsafeAddr arr[0]
  globalThingsLen = arr.len

proc dosomething() =
  let arr = cast[ptr UncheckedArray[Thing]](globalThings)
  for i in 0..<globalThingsLen:
    echo arr[i]

proc main =
  var myArrayOfThings: array[3, Thing]
  myArrayOfThings[0] = Thing(name: "alice")
  myArrayOfThings[1] = Thing(name: "bobby")
  myArrayOfThings[2] = Thing(name: "cathy")
  save myArrayOfThings
  myArrayOfThings[1] = Thing(name: "A walrus")
  dosomething()

when isMainModule:
  main()
```

For me, that produces the following when compiled with `--gc:orc`, which is what I'd hope it would produce:

```
(name: "alice")
(name: "bobby")
(name: "cathy")
```

Reply:
orc will only free stuff when they get out of scope, so as soon as main is finished, `myArrayOfThings` will no longer exist.
You can check by calling `dosomething()` after calling `main()`

Reply:
`trustAnchors` should be put on the heap. You might add a field of the `seq[X509TrustAnchor]` type to `TLSAsyncStream` so that it gets the same lifetime as `ctx`.

Reply:
That's what I was thinking, but it seems like it could be an awful waste of memory to duplicate it for every stream compared with the current implementation that shares the anchors between streams. In my particular use case, I don't mind making a copy per stream (as I'll only have one trust anchor and very few client streams) but don't want to ruin the more common case in the process.

I suppose I could make it use a `seq[X509TrustAnchor]` if provided, otherwise use the const Mozilla anchors as-is?

Reply:
List of trusted anchors is static array which should not be changed while process is running, `seq[X509TrustAnchor]` unable to satisfy this requirement, its mutable data structure. And i understand that we should somehow preserve this array in memory because of GC but storing it with every instance of `TlsAsyncStream` is also not a good solution. So i prefer to have `openArray[X509TrustAnchor]` and some comments in function description that you should be sure that specified trusted anchors list will be preserved in memory while `TlsAsyncStream` is alive.